corehq/apps/reports/util.py

```diff
@@ -407,3 +407,92 @@ def from_request_dict(cls, query):
         search = query.get("sSearch", "")
 
         return DatatablesParams(count, start, desc, echo, search)
+
+
+# --- Tableau API util methods ---
+
+TableauGroupTuple = namedtuple('TableauGroupTuple', ['name', 'id'])
+
+
+def _group_json_to_tuples(group_json):
+    group_tuples = [TableauGroupTuple(group_dict['name'], group_dict['id']) for group_dict in group_json]
+    # Remove default Tableau group:
+    for group in group_tuples:
+        if group.name == 'All Users':
+            group_tuples.remove(group)
+            break
+    return group_tuples
+
+
+@quickcache(['domain'], timeout=30 * 60)
+def get_all_tableau_groups(domain):
+    '''
+    Returns a list of all Tableau groups on the site as list of TableauGroupTuples.
+    '''
+    session = TableauAPISession.create_session_for_domain(domain)
```

Comment:
These methods create a session and sign in but they never sign out. Is that OK? Making the session a [context manager](https://docs.python.org/3/library/stdtypes.html#typecontextmanager) would provide clear delineation for where the session is expected to be active and it would also separate the creation of it from signing in / out:

```python
# create session (but it's not signed in yet)
session = TableauAPISession.create_session_for_domain(domain)
with session:  # this signs in
    # do something

# closing the context manager could sign out if we want that - or at least clear the token from memory if we don't care about signing out.
```

Reply:
Let's stick with JWT rather than personal tokens. Adding a comment about signing out being broken would be useful. I think it's probably fine not to sign out since we never save the token anywhere but reducing the token TTL does sound like a good idea.